scrapy/spiders/crawl.py

```diff
@@ -74,48 +81,61 @@ def parse_start_url(self, response):
     def process_results(self, response, results):
         return results
 
-    def _build_request(self, rule, link):
-        r = Request(url=link.url, callback=self._response_downloaded)
-        r.meta.update(rule=rule, link_text=link.text)
-        return r
+    def _build_request(self, rule_index, link):
+        return Request(
+            url=link.url,
+            callback=self._callback,
+            errback=self._errback,
+            meta=dict(rule=rule_index, link_text=link.text),
+        )
 
     def _requests_to_follow(self, response):
         if not isinstance(response, HtmlResponse):
             return
         seen = set()
-        for n, rule in enumerate(self._rules):
-            links = [lnk for lnk in rule.link_extractor.extract_links(response)
-                     if lnk not in seen]
-            if links and rule.process_links:
-                links = rule.process_links(links)
-            for link in links:
+        for rule_index, rule in enumerate(self._rules):
+            links = (lnk for lnk in rule.link_extractor.extract_links(response)
+                     if lnk not in seen)
+            for link in rule.process_links(links):
```

Comment:
I think it is better to keep links as lists, because it was documented  that Rule.process_links input is a list (see  process_links docs).

Reply:
Makes sense. I'll revert this change, and will open a follow-up PR including this, [this comment](https://github.com/scrapy/scrapy/pull/4000#discussion_r321303103) and some other documentations improvements. Thanks!

Reply:
@kmike Updated, please check again. Thanks!